### ISSUE TITLE
Rearchitect downloader design

### ISSUE THREAD

**Issue by mike-burrage-hedera** [hashgraph/hedera-mirror-node#314]
**Detailed Description**
The downloader component is inefficient and needs to be rewritten to:

- Maximize the download throughput from the cloud (signatures _and_ record/balance files)
  - Use some sort of priority queue to schedule downloads
- Minimize tasks done serially in the downloader
  - currently AWS SDK v1 performs several serial HTTP requests prior to queueing the actual file download in the TransferManager
  - non-signature files are currently only downloaded after enough signatures have been downloaded for that record/event/balance stream (the actual download can be done in parallel without actually processing it or adding it to the DB until enough signatures have been verified)
- Store the files and process them in memory as opposed to on the filesystem (eliminating the use of the file system as a signal between the downloader and record processor).
- Potentially download from several cloud sources in parallel if that helps (google cloud and AWS S3 in parallel) - if downloading from a single source becomes a bottleneck
- Separate the logic that determines what files are available in the cloud (currently via polling the bucket) from the actual priority queue/download piece. We may have better alternatives to constantly polling the bucket (receiving notifications from the bucket instead of polling).

We should also consider how we want to handle "catch up" processing for standing up a clean/new node with an empty database but a long history of files in the cloud. Possibly involving prioritization of the downloads (always try to download the most recent files when available, and catch up with older files at a lower priority) - this requires more discussion.

**Actual Behavior**
- Currently using AWS sdk v1 TransferManager.
- record and account balance files (not including signatures) are serially downloaded.

**Expected Behavior**

**Environment:**

**Additional Context**
AWS SDK v2 has many improvements over v1. Downloading can be done on a single thread/event loop - eliminating the requirement for an additional thread per download (everything is IO-bound here). This is an option.

**Comment by apeksharma** [comment]
Doc: https://hederatest.sharepoint.com/:w:/s/ProductHome/ESnqAJDgBLpCiv4PATddShcBho_a6SBMdHYAE2xgczebIA?e=bWX9Qh

**Comment by apeksharma** [comment]
Doc: https://hederatest.sharepoint.com/:w:/s/ProductHome/ESnqAJDgBLpCiv4PATddShcBho_a6SBMdHYAE2xgczebIA?e=bWX9Qh

**Comment by steven-sheehy** [comment]
Closing as too broad
